### PR TITLE
Two fixes

### PR DESCRIPTION
**Pull request by axefor**
This PR fixes two smaller bugs.

1. If the Scipy minimize optimizer fails, it might leave a stale loss state from optimization trials, while the latest "result" that is actually returned might still be usefull. At failiure, the loss is therefore recomputed with the returned parameters.

2. In the stress loss function part, the volumes have been computed for `self.idcs_str`. However, they are later refered as `self.volumes[i]`, where `i` runs over all indices. (`if i not in self.idcs_str: continue` skipps non–stress containing indices.) The fix is to make a list of all volumes.

Edit: Make 1. rely on #125 and always call loss in ParallelOptimizerBase.optimize, so that this fixes all optimizers. #125 still avoids recalculation when not needed.

### CODE REVIEW

**Review by yuzie007 — APPROVED** [review]
Thank you for the fix! I also personally wonder if the `_worker_loop` update might solve my concern about the "dead lock" I observed in training in parallel CPUs. After this is merged, I will test it again.
